### ISSUE TITLE
re2 is two times slower than re in this scenario

### ISSUE THREAD

**Issue by d4rk-d4nph3** [andreasvc/pyre2#12]
## Environment:
* Python 3.7.7 on MacOS 10.15.3(Catalina)
* Fresh install of re2.

## Scenario:
I have a blacklist file of IPs from where I want to extract all the IPs to a list for my use case. The list contains roughly 32K IPs and contains a variable header. I extracted all the IPs using a simple regex for IP where I am sure that all the IPs are valid.

## Code:
```python
import re2

re_ip_search = re2.compile(r'\d+\.\d+\.\d+\.\d+')
blacklist_file = 'bi_ssh_0_1d.ipset'

with open(blacklist_file, 'r') as file:
    lines = [line.rstrip('\n') for line in file]

ip_list = list(filter(re_ip_search.match, lines))
```
## Benchmark:
I used timeit to benchmark the performance:
```python
from timeit import timeit
exec_time = timeit('list(filter(re_ip_search.match, lines))', 
                    number=1, setup = 'from __main__ import re_ip_search, lines')

print(exec_time*1000)   
```
## Result
re  ->  14.20 ms (avg)
re2 -> 28.6 ms (avg)

Why is re2 slower by two times in this case. Am I missing something?

**Comment by andreasvc** [comment]
I don't have data to run your benchmark at hand, so I'll just mention the first things that come to mind:

1. maybe creating the Match object is what's expensive for re2; in this case they are not needed. maybe a match method that returns a bool should be added.
2. calling match on each line separately comes with a lot of function call overhead. using findall() on the whole file is probably faster.
3. re2 has to convert unicode to utf8; this can be avoided by reading the file in binary mode as bytes.

**Comment by d4rk-d4nph3** [comment]
Thanks for the quick reply. I did what you advised by using _findall_ to the whole file instead of per line by match. 
```python
import re2

&#35; Refactored code
re_ip_search = re2.compile(r'\d+\.\d+\.\d+\.\d+')
inp_file = 'bi_ssh_0_1d.ipset'

with open(inp_file, 'r') as file:
    lines = re_ip_search.findall(file.read())

```
This worked like a charm with now, the performance of re2 is on par with re. Further, reading file in binary mode and using byte regex slightly improved the performance by about 1-2 ms. 

### Conclusion
It seems to be that re2 and re perform equal in this case due to the type of regex used and sometimes we have to think outside the box and not just plugging re2 as a drop in replacement for re.

**Comment by d4rk-d4nph3** [comment]
One last thing. Is this library production stable? Looking to hook this up after upgrading to python 3.

**Comment by andreasvc** [comment]
As far as I know it's pretty good. There are tests and it has worked well in my academic experiments, but your mileage may vary.

I replicated your benchmark: https://gist.github.com/andreasvc/aba3a5055ec780601a7e2ef83076f08d
I'm using a bytes pattern on bytes data with a mix of valid and invalid IPs. I included the new contains method, but re2.findall is the fastest:
```
re match 12.2
re findall 14.1
re2 match 17.3
re2 contains 10.4
re2 findall 5.7
re2 count 5.7
```
However, if all lines are valid IP addresses, re.findall and re2.findall perform similarly, which is interesting.

I think re2.match could be sped up by delaying the creation of the spans and groups until they are needed, but I don't know if it's worth the effort because it's not a trivial change.
